plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/SourceRole.kt

```diff
@@ -30,15 +28,18 @@ sealed interface SourceRole {
             extractInfoFromFunctionArgument(0).unwrap<FirSymbolHolder>().firSymbol
     }
 
-    data class ConditionalEffect(val lhs: SourceRole?, val rhs: SourceRole?) : SourceRole
-    data class IsTypeCondition(val targetVariable: FirBasedSymbol<*>, val expectedType: ConeKotlinType, val negated: Boolean = false) :
-        SourceRole
+    data class CallsInPlaceEffect(val paramSymbol: FirBasedSymbol<*>, val kind: EventOccurrencesRange) : SourceRole
+    data class ConditionalEffect(val effect: SourceRole, val condition: Condition) : SourceRole
+    data class FirSymbolHolder(val firSymbol: FirBasedSymbol<*>) : SourceRole, Condition
```

Comment:
As on the right-hand-side of an implication there could be an instance of a boolean variable, also `FirSymbolHolder` is a `Condition`. 

```kotlin
fun test(b: Boolean): Boolean {
    contract {
        returns(true) implies b  
    }
    return true
}
```

The function `ContractDescriptionConversionVisitor::visitValueParameterReference` returns a variable embedding. This embedding is an instance of a `FirVariableEmbedding`. Following pull-request #184, `FirVariableEmbedding`s have a source role (`FirSymbolHolder`) on them.

Reply:
Perhaps it would be better to wrap the `FirSymbolHolder` into a dedicated condition so that its role is clear?

I would make `Condition` a subinterface of `SourceRole`.

Reply:
> Perhaps it would be better to wrap the FirSymbolHolder into a dedicated condition so that its role is clear?

Yeah, I agree with your point, but then I should find a way to override the `sourceRole` property of a `FirVariableEmbedding`.

> I would make Condition a subinterface of SourceRole.
Ok!

Reply:
> Yeah, I agree with your point, but then I should find a way to override the sourceRole property of a FirVariableEmbedding.

Ah, I see.  Fair, maybe not worth the trouble at the moment then.